OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -70,245 +71,177 @@ public class WeatherOverlayInfo : ITraitInfo, ILobbyCustomRulesIgnore
-					tempParticleCount = maxParticleCount - particleList.Count;
+				Pos = pos;
+			}
 
-				SpawnParticles(tempParticleCount, (int)tempRangeY, 50);
+			public Particle(Particle source, float2 pos, int swingDirection, float swingOffset)
+				: this(source)
+			{
+				Pos = pos;
+				SwingDirection = swingDirection;
+				SwingOffset = swingOffset;
 			}
 		}
 
-		void FadeOutParticleCount(WorldRenderer wr)
+		readonly WeatherOverlayInfo info;
+		readonly World world;
+
+		float windStrength;
+		int targetWindStrengthIndex;
+		long windUpdateCountdown;
+		Particle[] particles;
+		Size viewportSize;
+
+		public WeatherOverlay(World world, WeatherOverlayInfo info)
 		{
-			for (var i = 0; i < particleList.Count; i++)
-				if (particleList[i].PosY > (Game.Renderer.Resolution.Height - particleList[i].Gravity))
-					particleList.RemoveAt(i);
+			this.info = info;
+			this.world = world;
+			targetWindStrengthIndex = info.ChangingWindLevel ? info.WindLevels.Length / 2 : 0;
+			windUpdateCountdown = world.LocalRandom.Next(info.WindTick[0], info.WindTick[1]);
 		}
 
-		void XAxisSwing(ref Particle tempParticle)
+		void INotifyViewportZoomExtentsChanged.ViewportZoomExtentsChanged(float minZoom, float maxZoom)
 		{
-			// Direction turn
-			if (tempParticle.SwingOffset < -tempParticle.SwingAmplitude || tempParticle.SwingOffset > tempParticle.SwingAmplitude)
-				tempParticle.SwingDirection *= -1;
+			// Track particles in a viewport fixed to the minimum zoom level
+			var s = (1f / minZoom * new float2(Game.Renderer.NativeResolution)).ToInt2();
+			viewportSize = new Size(s.X, s.Y);
 
-			// Perform the X-Axis-Swing
-			tempParticle.SwingOffset += tempParticle.SwingDirection * tempParticle.SwingSpeed;
+			// Randomly distribute particles within the initial viewport
+			var particleCount = viewportSize.Width * viewportSize.Height * info.ParticleDensityFactor / 10000;
+			particles = new Particle[particleCount];
+			var rect = new Rectangle(int2.Zero, viewportSize);
+			for (var i = 0; i < particles.Length; i++)
+				particles[i] = new Particle(info, world.LocalRandom, rect);
 		}
 
 		void ITick.Tick(Actor self)
-		{
-			windTickCountdown--;
-		}
-
-		void WindLogic(ref Particle tempParticle)
 		{
 			if (!info.ChangingWindLevel)
-				targetWindXOffset = info.WindLevels[0];
-			else if (windTickCountdown <= 0)
+				return;
+
+			if (--windUpdateCountdown <= 0)
 			{
-				windTickCountdown = Game.CosmeticRandom.Next(info.WindTick[0], info.WindTick[1]);
-				if (Game.CosmeticRandom.Next(2) == 1 && currentWindIndex > 0)
-				{
-					currentWindIndex--;
-					targetWindXOffset = info.WindLevels[currentWindIndex];
-				}
-				else if (currentWindIndex < info.WindLevels.Length - 1)
-				{
-					currentWindIndex++;
-					targetWindXOffset = info.WindLevels[currentWindIndex];
-				}
+				windUpdateCountdown = self.World.LocalRandom.Next(info.WindTick[0], info.WindTick[1]);
+				if (self.World.LocalRandom.Next(2) == 1 && targetWindStrengthIndex > 0)
```

Comment:
While we're here it might be a good idea to change this to `if (targetWindStrengthIndex > 0 && self.World.LocalRandom.Next(2) == 1)` so we don't calculate unnecessary random values.